### ISSUE TITLE
java.lang.IllegalStateException: Flusher when using HTTP/3 with Spring Boot 3.2

### ISSUE THREAD

**Issue by murphye** [eclipse/jetty.project#10519]
**Jetty version(s)**
Jetty 12.0.1

**Java version/vendor** `(use: java -version)`
openjdk version "20.0.1" 2023-04-18

**OS type/version**
macOS 13.5.2

**Description**

I have created a functional HTTP/3 Spring Boot 3.2 demo using Jetty 12. I have detailed documentation in the README on how to build and run the demo application: https://github.com/murphye/spring-boot-http-3-jetty

You can see in https://github.com/murphye/spring-boot-http-3-jetty/blob/main/src/main/java/com/example/demo/JettyConfiguration.java how I configure `HTTP3ServerConnector`.

Randomly, I am getting an exception from Jetty. It does not happen very often and occurs randomly as far as I can tell. Here is the stack trace:

```
java.lang.IllegalStateException: Flusher@4df6445d[CLOSED]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:220) ~[jetty-util-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.common.QuicSession.flush(QuicSession.java:363) ~[jetty-quic-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.server.ServerQuicSession.flush(ServerQuicSession.java:126) ~[jetty-quic-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.common.QuicSession.outwardClose(QuicSession.java:405) ~[jetty-quic-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.common.ProtocolSession.outwardClose(ProtocolSession.java:160) ~[jetty-quic-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.HTTP3Session.outwardClose(HTTP3Session.java:692) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.HTTP3Session.outwardDisconnect(HTTP3Session.java:672) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.HTTP3Session.lambda$inwardClose$12(HTTP3Session.java:652) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.util.Callback$4.completed(Callback.java:211) ~[jetty-util-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.util.Callback$Completing.succeeded(Callback.java:378) ~[jetty-util-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.ControlFlusher.lambda$succeeded$1(ControlFlusher.java:118) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) ~[na:na]
        at org.eclipse.jetty.http3.ControlFlusher.succeeded(ControlFlusher.java:118) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:291) ~[jetty-io-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:254) ~[jetty-io-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:374) ~[jetty-io-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.ControlFlusher.process(ControlFlusher.java:106) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:243) ~[jetty-util-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[jetty-util-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.server.internal.ServerHTTP3Session.writeControlFrame(ServerHTTP3Session.java:274) ~[jetty-http3-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.server.internal.HTTP3SessionServer.writeControlFrame(HTTP3SessionServer.java:91) ~[jetty-http3-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.HTTP3Session.inwardClose(HTTP3Session.java:649) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.HTTP3Session.onIdleTimeout(HTTP3Session.java:602) ~[jetty-http3-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.http3.server.internal.ServerHTTP3Session.onIdleTimeout(ServerHTTP3Session.java:233) ~[jetty-http3-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.common.QuicSession.onIdleTimeout(QuicSession.java:204) ~[jetty-quic-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.server.ServerQuicSession.onIdleTimeout(ServerQuicSession.java:109) ~[jetty-quic-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.server.ServerQuicConnection$SessionTimeouts.onExpired(ServerQuicConnection.java:136) ~[jetty-quic-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.server.ServerQuicConnection$SessionTimeouts.onExpired(ServerQuicConnection.java:118) ~[jetty-quic-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.io.CyclicTimeouts.onTimeoutExpired(CyclicTimeouts.java:113) ~[jetty-io-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.io.CyclicTimeouts$Timeouts.onTimeoutExpired(CyclicTimeouts.java:206) ~[jetty-io-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.io.CyclicTimeout$Wakeup.run(CyclicTimeout.java:294) ~[jetty-io-12.0.1.jar:12.0.1]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577) ~[na:na]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[na:na]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:1623) ~[na:na]
```

**How to reproduce?**

Clone https://github.com/murphye/spring-boot-http-3-jetty to your local machine. Read through the README to understand the requirements (you can skip curl). I use a Mac but for the Firefox demo, Linux should work. The project is setup to use Java 20.

```shell
./mvnw clean compile spring-boot:run
```

Open Firefox and go to https://demo.local:8443. Reload and it will switch to HTTP/3 (H3) mode (see the README). 

Keep reloading until the `java.lang.IllegalStateException: Flusher@4df6445d[CLOSED]` exception occurs. 

Try to open multiple private windows in Firefox and also try using `curl --http3` which may also help trigger this.

This could take several minutes to reproduce. It may be best to do some reloads, stop, and try again in a few minutes. This seems to randomly occur but does occur consistently when it runs long enough.

Note: I am running H3 on 8443 along with H2. This problem also occurred when running H3 on a different port.

Note: This may happen more often when there are multiple clients (i.e. Firefox and curl) or just using curl

**Comment by sbordet** [comment]
@murphye that's great 🎉 

Some random comment:

* You are using a `Filter` to add the `Alt-Svc` header, but this should already be done by `HTTP3ServerConnectionFactory`, which adds a customizer that adds `Alt-Svc`. However, the value of the header is different. Did you find the additional attributes necessary?

* The stack trace seems to be due to idle timeout. What did you configure as idle timeout value on the connector? I think the stack trace is just a little too verbose, but should not really be a problem.

Thanks!

**Comment by murphye** [comment]
> You are using a Filter to add the Alt-Svc header, but this should already be done by HTTP3ServerConnectionFactory, which adds a customizer that adds Alt-Svc. However, the value of the header is different. Did you find the additional attributes necessary?

I am not seeing any `Alt-Svc` response header when I disable my `AddResponseHeaderFilter`. Could you please take a look at my code to see if I am missing anything to automatically create the `Alt-Svc` on the initial HTTP/2 response? https://github.com/murphye/spring-boot-http-3-jetty/blob/main/src/main/java/com/example/demo/JettyConfiguration.java

> The stack trace seems to be due to idle timeout. What did you configure as idle timeout value on the connector? I think the stack trace is just a little too verbose, but should not really be a problem.

I added `httpConfig.setIdleTimeout(idleTimeout);` with a value of `900000` (15 mins) via `application.properties`. I am still getting the same stack trace, and it may be related to the idle timeout.  If there is not a problem here, then there should not be a stack trace?

```
java.lang.IllegalStateException: Flusher@1642bd3c[CLOSED]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:220) ~[jetty-util-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.common.QuicSession.flush(QuicSession.java:363) ~[jetty-quic-common-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.server.ServerQuicSession.flush(ServerQuicSession.java:126) ~[jetty-quic-server-12.0.1.jar:12.0.1]
        at org.eclipse.jetty.quic.common.QuicSession.outwardClose(QuicSession.java:405) ~[jetty-quic-common-12.0.1.jar:12.0.1]
...
```

Thank you for any help that you can provide to button this down. I intend to write a blog post on this in the near future.

**Comment by murphye** [comment]
@sbordet For the issue related to `Alt-Svc`, I ran the debugger and in the following screenshot, the first breakpoint is hit, but not the second (when issuing a HTTP/2 request). What could cause this? The `HTTP3ServerConnector` is definitely working otherwise.

The code in https://github.com/murphye/spring-boot-http-3-jetty/blob/main/src/main/java/com/example/demo/JettyConfiguration.java is close to what you have in the docs https://eclipse.dev/jetty/documentation/jetty-11/programming-guide/index.html#pg-server-http-connector-protocol-http3

<img width="927" alt="image" src="https://github.com/eclipse/jetty.project/assets/643097/11f0035f-46be-4a66-9b18-02d5aa8cc003">

**Comment by lorban** [comment]
@murphye A fix for this problem just got merged into the jetty-12.0.x branch. Could you please confirm that it works for you by either building the branch yourself (or waiting until tomorrow for the nightly build to be published in the [Jetty snapshot repo](https://oss.sonatype.org/content/repositories/jetty-snapshots)) and report back here?

Thanks!
